### PR TITLE
klient: block remote methods on unpaid subscription

### PR DESCRIPTION
**Pull request by ppknap**
Remote auto-updated clients will refuse callers' request when klient's team is blocked due to unpaid subscription.

Depends on: ~#9870~

| Method   |      Callers      |  Blocked | Command |
|----------|:-------------:|:------:|:-------:|
| `klient.usage` |  kloud | :x: | - |
| `os.home` |    klient   | :white_check_mark: | - | 
| `os.currentUsername` | klient | :white_check_mark: | ssh |
| `klient.info` | klient | :white_check_mark: | list |
| `klient.disable` | browser | :x: | - |
| `klient.share` | kloud/browser/socialapi | :x: | - |
| `klient.unshare` | kloud/browser/socialapi | :x: | - |
| `klient.shared` | browser/socialapi | :x: | - |
| `sshkeys.list` | - | :white_check_mark: | - |
| `sshkeys.add` | klient | :white_check_mark: | ssh |
| `sshkeys.delete` | - | :white_check_mark: | - |
| `storage.set` | browser | :x: | - |
| `storage.get` | browser | :x: | - |
| `storage.delete` | browser | :x: | - |
| `log.tail` | browser | :x: | - |
| `fs.readDirectory` | browser/klient | :white_check_mark: | mount |
| `fs.glob` | browser/klient | :white_check_mark: | mount |
| `fs.readFile` | browser/klient | :white_check_mark: | mount |
| `fs.writeFile` | browser/klient | :white_check_mark: | mount |
| `fs.uniquePath` | browser/klient | :white_check_mark: | mount |
| `fs.getInfo` | browser/klient | :white_check_mark: | mount |
| `fs.setPermissions` | browser/klient | :white_check_mark: | mount |
| `fs.remove` | broser/klient | :white_check_mark: | mount |
| `fs.rename` | browser/klient | :white_check_mark: | mount |
| `fs.createDirectory` | browser/klient | :white_check_mark: | mount |
| `fs.move` | browser/klient | :white_check_mark: | mount |
| `fs.copy` | klient | :white_check_mark: | mount |
| `fs.getDiskInfo` | klient | :white_check_mark: | mount |
| `fs.getPathSize` | klient | :white_check_mark: | mount |
| `vagrant.create` | kloud | :x: | - |
| `vagrant.provider` | kloud | :x: | - |
| `vagrant.list` | kloud | :x: | - |
| `vagrant.up` | kloud | :x: | - |
| `vagrant.halt` | kloud | :x: | - |
| `vagrant.destroy` | kloud | :x: | - |
| `vagrant.status` | kloud | :x: | - |
| `vagrant.version` | kloud | :x: | - |
| `vagrant.listForwardedPorts` | klient | :x: | - |
| `tunnel.info` | kloud | :x: | - |
| `log.upload` | - | :x: | - |
| `webterm.getSessions` | browser | :white_check_mark: | - |
| `webterm.connect`  | browser | :white_check_mark: | - |
| `webterm.killSession` | browser | :white_check_mark: | - |
| `webterm.killSessions` | browser | :white_check_mark: | - |
| `webterm.rename` | browser | :white_check_mark: | - |
| `client.Publish` | KD | :x: | - |
| `client.Subscribe` | browser | :x: | - |
| `client.Unsubscribe` | browser | :x: | - |
## Motivation and Context
Payment handling

## How Has This Been Tested?
Manually

### CODE REVIEW

File: go/src/koding/klient/app/klient.go [hunk 1]
```diff
@@ -550,6 +554,26 @@ func (k *Klient) RegisterMethods() {
...
 }
 
+// handleWithSub is a middle-ware function that checks team payment status
+// before invoking fn function. It will fail if team is blocked due to unpaid
+// subscription.
+func (k *Klient) handleWithSub(method string, fn kite.HandlerFunc) {
+	k.kite.HandleFunc(method, func(r *kite.Request) (interface{}, error) {
+		team, err := k.cacheTeam()
+		if err != nil {
+			k.log.Error("Cannot find Klient's team: %s", err)
```
Comment (rjeczalik) [inline go/src/koding/klient/app/klient.go]: Won't work for KD for a reason stated in previous PR.

File: go/src/koding/klient/app/klient.go [hunk 2]
```diff
@@ -461,11 +465,11 @@ func (k *Klient) RegisterMethods() {
 	k.kite.HandleFunc("klient.usage", k.usage.Current)
 
 	// klient os method(s)
-	k.kite.HandleFunc("os.home", kos.Home)
-	k.kite.HandleFunc("os.currentUsername", kos.CurrentUsername)
+	k.handleWithSub("os.home", kos.Home)
```
Comment (rjeczalik) [inline go/src/koding/klient/app/klient.go]: How about making this more configurable?

Create a configuration map [like this one](https://github.com/koding/koding/pull/9882/commits/1389cf689dcee7f2cd0d1fed465fbe33a7e82bb1#diff-514e03b9d69e198098d3560e3bda3763R213), and register a PreHandlerFunc that if sub-check is requested, will validate it and return error if sub is not valid. Basically would work the same as auth check etc.

Comment (ppknap) [inline go/src/koding/klient/app/klient.go]: your link is broken

Comment (rjeczalik) [inline go/src/koding/klient/app/klient.go]: > If yes, why didn't you use proposed solution in your presence logic?

The idea just popped to my head why I was reading this klient.go file, remote.go was pretty raw. If you don't see a benefit then there's no need for changes. :-)

Comment (ppknap) [inline go/src/koding/klient/app/klient.go]: Honestly speaking, the current form was based on your solution for `remote` methods and I'm not against other ideas. Personally, I would have written something like:

```go
package subscription

type Check struct {
    Team func() (team.Team error) // team getter.
}

func (c *Check) Check() {
  // business logic here
}

// ...

func KiteCheck(c *Check, h kite.Handler) kite.Handler {
  return kite.HandlerFunc(func(req *Request) (result interface{}, err error) {
    if err := c.Check(); err != nil { // validation logic
      return nil, err
    }
    return h(req)
  })
}

```
and 
```go
k.kite.Handle("os.currentUsername", subscription.KiteCheck(c, kite.HandlerFunc(kos.CurrentUsername)))
```
With this, I could test the logic separately from klient/kite. But, I know that I sometimes overdo with my architecture ideas so I'm trying to not push them on anyone. :)

**Review by rjeczalik — APPROVED** [review]

File: go/src/koding/kites/kloud/stack/whoami.go [hunk 3]
```diff
@@ -16,6 +16,16 @@ type Team struct {
 	Subscription string `json:"subscription,omitempty"`
 }
 
+// IsSubActive checks if Team's subscription is active.
+func (t *Team) IsSubActive() bool {
```
Comment (cihangir) [inline go/src/koding/kites/kloud/stack/whoami.go]: lets not duplicate the logic, and have the critical logic in one place, please refactor this and the one we have in group.go to use the exact same code.

**Review by cihangir — CHANGES_REQUESTED** [review]

**Comment by ppknap** [comment]
@cihangir done

**Review by cihangir — APPROVED** [review]

**Comment by cihangir** [comment]
go build is failing, could you do repo-wide `s/PaymentStatusActive/SubStatusActive` ? You can merge when tests passes

**Review by rjeczalik — APPROVED** [review]

**Pull request merged**